Add spektrafilm module documentation (#993)
darktable-org/dtdocs: content/module-reference/processing-modules/spektrafilm.md

---
title: spektrafilm
id: spektrafilm
weight: 10
include_toc: true
---

{{< details summary="Technical information" class="technical-info" >}}

description

> : simulates the physical process of developing and printing analog film, using spectral emulsion and paper data from the [spektrafilm](https://github.com/andreavolpato/spektrafilm) project.

Reviewer: I'd move the spektrafilm link into the first paragraph to give this reference more visibility. The "technical information" block is collapsed by default

Because the underlying data is spectral rather than a fixed RGB transform, the module reacts to your working color space and white balance the way a real film stock would react to different light: the same film and paper combination will render differently under different scene illuminants, just as it would in a darkroom.

---

**Note**: Like other view transforms, modules placed before _spektrafilm_ in the pipeline operate in [scene-referred](../../../darkroom/pixelpipe/the-pixelpipe-and-module-order.md/#scene-referred-workflow) space. Modules after it work in [display-referred](../../../darkroom/pixelpipe/the-pixelpipe-and-module-order.md/#display-referred-workflow) space.

---

# usage

only use one display transform

> : Never use _spektrafilm_ together with another display transform module (i.e. [_filmic rgb_](./filmic-rgb.md), [_sigmoid_](./sigmoid.md), [_AgX_](./agx.md) or [_base curve_](./base-curve.md)) -- _spektrafilm_ performs the film's own tone mapping as part of simulating development and printing.

Reviewer: I find never quite harsh. Nothing catastrophic happens by combining tonemappers. Maybe: "It is not recommended...".

Reviewer 2: @anoderay The Sigmoid module description contains similarly worded advice:
"Never use sigmoid together with another display transform module (i.e. filmic rgb, AgX or base curve)."

Of course, any creative _distortion_ of an image has a right to exist, including those created using more than one tone mapper. But applying a linear tone mapper to nonlinear data results in "burned-out" highlights, unnatural color distortion, and loss of shadow detail.

Author: I would also prefer to lease as is here

start simple
: Selecting a film stock and a print paper is enough to get a complete, physically-plausible result. The per-effect controls (grain, halation, diffusion) and the film tab's _chemistry_ section are there for fine-tuning, not required for a first pass.

positive film has no print stage
: Slide and reversal film stocks are viewed directly rather than printed -- _scan the film_ is automatically enabled when you select one, and every print-stage control (print exposure compensation, auto print exposure, print contrast, print development time, filtration, preflash, print diffusion) has no effect while it's active. The scanner tab's _viewing glare_ is also inactive, since a directly scanned film has no print surface.

auto print exposure is a real-world default, not a fixed brightness

> : With _auto print exposure_ enabled, changing _film exposure_ does not change the final brightness -- print exposure automatically compensates, the same way a real printer exposes to a fixed target density regardless of how the negative was exposed. Disable it if you want film exposure to affect brightness directly, as it would with a fixed enlarger exposure time.

Reviewer: Needs update per the new default I think?

narrow-spectral-sensitivity papers may need manual print exposure

> : A few print stocks (duplicating films in particular) have a very narrow spectral sensitivity. _auto print exposure_'s metering can under- or over-shoot for these -- if a print looks implausibly dark or bright with auto print exposure enabled, use _print exposure compensation_ to correct it manually.

Reviewer: Maybe add examples?

# module controls

> Film stock, print paper, and the film format are always visible at the top of the module. Everything else is organized into tabs matching the stages of the physical process: _film_, _print_, _grain_, _halation_, _diffusion_, and _scanner_.

Reviewer: Maybe substitute "stages" with "aspects" as e.g. grain is not really a "stage" per se.

### chemistry

development time
: The development time the film's characteristic curves were measured at, in minutes. Only black & white stocks are characterised at more than one development time (Kodak Double-X at 4/5/6.5/9/12 minutes, for example); the slider is greyed out for colour stocks and for stocks with a single characterisation. The value snaps to the nearest measured time, and 0 selects the stock's own standard development. Unlike _development gamma_ below, which morphs the curves mathematically, this selects a different set of actually-measured curves.

development gamma
: Overall development contrast, applied by morphing the film's own density curves -- extended or reduced development time, as in push/pull processing. 1.0 is normal development.

fast layer gamma
: Contrast of the fastest (most light-sensitive) emulsion sub-layer only, independent of the slow layer -- push/pull processing doesn't always affect every sub-layer equally.

slow layer gamma
: Contrast of the mid and slow emulsion sub-layers.

developer exhaustion

> : Local developer depletion in dense (highly-exposed) areas: blends the highlight shoulder toward a self-limiting rolloff without shifting mid-grey. 0 disables it.

Reviewer: From my experience this increases highlight contrast; maybe this fact could be added here. At least to me "blends the highlight shoulder toward a self limiting rolloff" is difficult to understand.

quality
: Trade-off between spectral accuracy and processing speed. Higher settings use a finer-resolution lookup table, interpolated with PCHIP splines and validated against the reference implementation.

## print

print exposure compensation
: Manual print brightness (enlarger exposure time), in EV. This is an offset either way: with _auto print exposure_ enabled it shifts the automatic result rather than being ignored -- see [usage](#usage) above.

auto print exposure
: Automatically compensate print exposure for film exposure changes, the way a real printer would expose to a fixed density regardless of the negative's own exposure. Has no effect while _scan the film_ is enabled, since there's no print stage to compensate.

print contrast
: Print contrast, applied by morphing the paper's own density curves rather than a simple RGB contrast operation.

### chemistry

> development time

Reviewer: Duplicate from above

: The filter family to simulate:
: - _black pro-mist_: a concentrated, punchy halo with deep blacks preserved.
: - _glimmerglass_: a tight, subtle diffusion that preserves sharpness.
: - _pro-mist_: broader and more pastel, a softer atmospheric look.
: - _cinebloom_: a frame-wide, slow-decaying veil.

diffusion strength
: Diffusion filter strength.

diffusion size
: Diffusion halo/bloom size.

diffusion halo warmth
: Warmth of the diffusion halo -- positive values warm the outer halo, negative values cool it. Added on top of the selected filter type's own inherent warmth bias.

### print diffusion

A second, independent diffusion filter applied at the print stage rather than the film stage -- simulating a filter placed at the enlarger instead of the camera. Its controls are the same as [diffusion](#diffusion) above, applied independently: _enable print diffusion_, _print diffusion filter type_, _print diffusion strength_, _print diffusion size_, _print diffusion halo warmth_.

## scanner

The scanner stage models how the developed film or finished print is digitised, and the conditions it is viewed under. These act on the final image, after everything else.

pre-compression boost

> : Multiplies XYZ luminance immediately before the OkLCh output gamut compressor, pushing the histogram to the right while the film's own shoulder rolloff is preserved. 1.0 is neutral/off. Because it acts at the very end of the module, the picker measures the processed image rather than the input: use it on an area of highlights to set the boost so the brightest tone lands just past the compressor's knee.

Reviewer: Could be less technical IMHO

purpose
: creative.

input
: linear, RGB, scene-referred.

processing
: non-linear, RGB.

output
: non-linear, RGB, display-referred.

{{< /details >}}

Recreate the look of a film stock printed onto darkroom paper.

Most film emulations are a colour recipe: someone photographed a scene on film, measured how the colours moved, and saved that as a curve or a LUT. This module works the other way round. It starts from laboratory measurements of what real film and real paper are made of -- how sensitive each colour layer is to each wavelength of light, and how much dye each one produces -- and simulates what physically happens to a photograph, step by step:

- light from the scene falls on the film and exposes its three colour layers,
- development turns that exposure into dye, with the layers chemically interfering with one another as they do in a real tank,
- an enlarger shines light through the developed negative onto a sheet of paper,
- the paper develops in turn, and the finished print is scanned.

> Grain, halation and diffusion filters are simulated as things that physically happen along the way, rather than as effects painted on at the end.

Reviewer 3: "X, rather than Y"

> Because the module works from what light actually is, the same film and paper render differently under different lighting -- exactly as real film does. Tungsten light and midday sun will not simply differ in white balance.

Reviewer: "Because the module works from what light actually is"

Nitpick if you allow: That sentence is a bit weird. I suggest:
`Because the module simulates the physical properties of light and its interaction with film exposure and printing process the same film and paper render differently under different lighting. Tungsten [...].`

Author: thanks!

> One consequence is worth knowing up front. Because the module works from what light actually is, rather than from a fixed colour recipe, the same film and paper render differently under different lighting -- exactly as real film does. Tungsten light and midday sun will not simply differ in white balance.

Reviewer 3: "X, rather than Y". Also the "One consequence …" is noise and "will not simply" could be replaced with a positive "more than" or something like that.

---

**Note**: Like other view transforms, modules placed before _spektrafilm_ in the pipeline operate in [scene-referred](../../../darkroom/pixelpipe/the-pixelpipe-and-module-order.md/#scene-referred-workflow) space. Modules after it work in [display-referred](../../../darkroom/pixelpipe/the-pixelpipe-and-module-order.md/#display-referred-workflow) space.

---

# usage

install the data first
: The module needs a data pack -- the film and paper measurements it works from -- and cannot render anything without one. If none is installed, the module shows a single button that downloads it, and the rest of the controls appear once that finishes. Packs are checked against a checksum before being installed, and are stored alongside your configuration, so they survive clearing the cache.

only use one display transform
: Never use _spektrafilm_ together with another display transform module (i.e. [_filmic rgb_](./filmic-rgb.md), [_sigmoid_](./sigmoid.md), [_AgX_](./agx.md) or [_base curve_](./base-curve.md)) -- _spektrafilm_ performs the film's own tone mapping as part of simulating development and printing.

start simple

> : Choosing a film stock is enough. Everything else already carries a sensible value taken from that film's own measurements, and the paper follows the film automatically. The tabs are for fine-tuning, not for getting a first result.

Reviewer 3: "X, not Y".

everything resets
: Double-click any slider to return it to its default. Each section heading has its own reset button for just that group of controls, and double-clicking a tab resets the whole tab -- useful for getting back to the film's own behaviour after experimenting, without starting over.

slide film has no print stage
: Slide and reversal stocks are viewed directly rather than printed. _scan the film_ switches on by itself when you choose one, and the entire print tab then has no effect, along with _viewing glare_ on the scanner tab -- there is no print surface for the light to reflect off.

auto print exposure changes what film exposure does

> : Out of the box, _film exposure_ behaves like leaving the enlarger on for a fixed time: expose the film more and the print comes out brighter. Switch on _auto print exposure_ and it stops doing that, because print exposure now compensates automatically -- which is what a real printer does, aiming for consistent print density however the negative was exposed. _film exposure_ then changes only colour and grain, by moving the scene to a different part of the film's response. It is off by default so that the control does the obvious thing until you ask for the darkroom behaviour.

Reviewer 3: Everything after "It is off by default" is awkward and could even be dropped.

# module controls

Film stock, print paper and film format are always visible at the top. Everything else is grouped into tabs, one per stage of the physical process: _film_, _print_, _grain_, _halation_, _diffusion_ and _scanner_.

## header

film stock
: The film to simulate.

print paper
: The paper to print onto. Left on _auto_ it follows the film stock's own intended paper, and names which one that currently is -- for example _auto (Kodak Portra Endura)_. Choose a paper explicitly and it stays put when you change film.

format
: A preset picker for common frame sizes (half-frame, 35mm, 6x6, 6x7, 6x9, 4x5, 8x10, Super 8, 16mm, Super 16, Super 35, VistaVision, 65mm 5-perf, IMAX 15-perf, or custom), which sets _frame long edge_ below. The preset names a film _gauge_ (35mm) while the slider gives the frame's long edge (36mm); both describe the same format.

frame long edge

> : The real-world width of the simulated frame, in mm. This is the physical scale everything else is measured against, so grain, scatter, halation and diffusion all come out proportionally larger on a smaller format at the same print size -- just as they do in reality. It is the honest way to make grain coarser: a smaller frame enlarged further shows bigger grain, exactly as it would on paper.

Reviewer 3: ```suggestion
: The real-world width of the simulated frame, in mm. This is the physical scale everything else is measured against, so grain, scatter, halation and diffusion all come out proportionally larger on a smaller format at the same print size -- just as they do in reality. It is the way to make grain coarser: a smaller frame enlarged further shows bigger grain, exactly as it would on paper.
```

## film

### exposure

film exposure
: Exposure adjustment at the film stage, in EV. With _auto print exposure_ on this no longer changes overall brightness (see [usage](#usage) above), but it still changes colour and grain, because it moves the scene to a different part of the film's response.

scan the film (skip print)
: Look at the developed negative or slide directly instead of printing it. Switches on by itself for slide and reversal stocks and off again for negative stocks; you can override it either way afterwards.

push/pull
: Shooting the film at a speed other than the box says, then compensating in development, in stops. Positive values push (shoot darker, develop longer), negative values pull. An approximation, since the real result depends on the specific developer, which is not simulated. Stacks with the _chemistry_ controls below.

### chemistry

development time

> : How long the film is developed, in minutes. Only a few black & white stocks were measured at more than one time -- Kodak Double-X at 4, 5, 6.5, 9 and 12 minutes, for instance -- so the slider is greyed out for colour films and for anything measured only once. Values snap to the times actually measured, and 0 means the stock's standard development. Unlike _development gamma_ below, this switches between genuinely different measurements rather than calculating something in between.

Reviewer 3: "genuinely" and "X rather than Y"

development gamma
: Overall development contrast, as though you had developed for longer or shorter. 1.0 is normal.

fast layer gamma
: Contrast of the most light-sensitive layer alone. Real film has several layers of differing sensitivity, and a change in development does not affect them equally.

slow layer gamma
: Contrast of the less sensitive layers.

developer exhaustion
: In a real developing tank the chemistry gets locally used up where the film was most heavily exposed. Those areas then stop gaining density however much brighter the scene gets, so highlights run into a ceiling instead of climbing indefinitely -- and the approach to that ceiling steepens, so highlight contrast tends to rise rather than soften. Mid-grey is held still, so only the bright end moves. 0 switches it off.

### DIR couplers

> As one colour layer develops, it releases a chemical that slows development in its neighbours. This is where much of colour negative film's saturation comes from, and it sharpens edges too: a bright area holds back development just beyond its own boundary, which reads as extra definition.

Reviewer 3: "This is where much" is Claudish.

DIR couplers

> : Overall strength of the effect. 1.0 matches the real film, 0 switches it off. The slider stops at 1.0 because the simulation has to be able to work backwards from the film's measured curves, and beyond film-accurate strength that becomes impossible -- for some stocks well before 2.0. The module quietly reduces the amount further for stocks where it would break down sooner.

Reviewer 3: ```suggestion
: Overall strength of the effect. 1.0 matches the real film, 0 switches it off. The slider stops at 1.0 because the simulation has to be able to work backwards from the film's measured curves, and beyond film-accurate strength that becomes impossible -- for some stocks well before 2.0. The module reduces the amount further for stocks where it would break down sooner.
```

same-layer inhibition
: How much each layer holds back its own development. Raising it flattens contrast within each colour channel.

interlayer inhibition
: How much each layer holds back the other two. This is the part that produces the saturation the effect is known for; at 0 the three colours develop independently.

inhibitor spread

> : How far the released chemical spreads through the emulsion, in thousandths of a millimetre. This is what turns the couplers into an edge effect rather than a purely tonal one: the further it spreads, the further a bright area reaches past its own boundary, and the more the result looks like added sharpening. Shorten it for a gentler, more purely tonal response -- 5 to 20 is the useful range. Each film carries its own value, and the slider follows the film when you change stock.

Reviewer 3: Another "This is what" Claudism.

spread tail

> : A small part of the chemical travels much further than the rest. This sets how far, and is what gives large bright areas a broad, gentle falloff rather than a glow that stops abruptly.

Reviewer 3: "This sets how far, and is what gives" is hard to comprehend.

tail weight

> : How much of the chemical travels in that long tail rather than spreading normally. 0 removes the tail entirely.

Reviewer 3: "X rather than Y"

### advanced

quality
: How finely the colour simulation is calculated. The three table settings work the answer out on a grid in advance and interpolate between the points, so a larger grid is closer to exact and slower to prepare. _exact spectral_ skips the grid and calculates every pixel directly: much slower, CPU only, and rarely visibly different.

bandwidth adaptation

> : Trims the extreme violet and deep red ends of the film's sensitivity, as part of how each stock is characterised. On by default and best left on -- it belongs to the film's description rather than being a look.

Reviewer 3: "X rather than Y" and in general everything after the pseudo em dash seems awkward here.

surface adaptation
: An optional per-colour exposure correction carried in the film data, worth up to two stops for strongly coloured light and nothing at all for neutral. Off by default, because it shifts saturated colours noticeably and the reference implementation does not apply it either. Stocks whose data does not include the correction are unaffected either way.

## print

### exposure

print exposure compensation
: How long the enlarger stays on, in EV -- the brightness of the print. Always an offset: with _auto print exposure_ on it shifts the automatic result rather than being ignored.

auto print exposure
: Compensate print exposure automatically for changes in film exposure, the way a real printer aims for consistent density whatever the negative. Does nothing while _scan the film_ is on.

print contrast

> : Contrast of the print, achieved by reshaping the paper's own response rather than by a generic contrast adjustment.

Reviewer 3: "X rather than Y"

### chemistry

development time
: How long the print is developed. Works exactly like the film tab's version and is set separately from it. Only Kodak Print Film 2302 was measured at more than one time (2, 3.5, 5, 7 and 9 minutes), so the slider is greyed out for every other paper -- and while _scan the film_ is on, since there is no print to develop.

### filtration

filtration M / filtration Y
: Magenta and yellow filters in the enlarger, in Kodak CC units away from neutral -- the colour balance of the print, adjusted the way it is in a darkroom.

### preflash

preflash exposure
: A brief, even pre-exposure of the paper before the main one, through the blank part of the film. Lifts shadows and lowers contrast; a standard darkroom technique for printing difficult negatives. 0 switches it off.

preflash M filter shift / preflash Y filter shift
: Magenta and yellow filtration for the preflash alone, in Kodak CC units, independent of the main filtration above.

## grain

> Film grain is not noise sprinkled over a sharp picture. Silver crystals in the emulsion develop or don't, individually, and the picture is made out of them -- so the simulation builds a grainy emulsion, blurs detail and grain together the way a real one does, and then restores the lost edge definition with the sharpening under _texture_ below. The blur and the recovery are tuned as a pair, so a picture with grain on is slightly softer than the same picture with grain off. That is how film behaves. If you want it sharper, turn grain down rather than pushing the recovery past its default.

Reviewer 3: The ", individually, " is awkward and also the first introduction sentence seems rather editorial. The "X rather than Y" is not that helpful either.

The controls are split in two. _emulsion_ changes what the film is made of, and rebuilds the crystals. _texture_ changes only how those crystals are drawn at your output size.

enable grain
: Switch grain simulation on.

### emulsion

granularity
: How coarse the film's crystals are, relative to the value measured for this stock. 1.0 is the datasheet figure.

> : This is the size control. Raising it grows the crystals, which means there are fewer of them and each one counts for more, so the grain becomes coarser *and* stronger together -- the same way a faster film differs from a slower one. Because it rebuilds the emulsion rather than adjusting the finished picture, it behaves the same on negative and slide film.

Reviewer 3: "X rather than Y"

grain strength
: How far the grain is allowed to move each pixel. 1.0 matches the real stock, 0 switches it off.

> : Unlike granularity this scales the result rather than the film, and that has one consequence worth knowing: a negative's grain gets amplified a second time by the print stage, while a slide is scanned directly with nothing to amplify it. The same value therefore reads much weaker on slide film. Reach for granularity rather than pushing this past 2 (right-click for up to 8).

Reviewer 3: "X rather than Y" plus the "and that has one consequence worth knowing" is Claude at 11.

uniformity

> : How evenly the crystals are distributed, relative to the stock's own figure. Lowering it bends the noise into a bell: grain that peaks in the midtones and eases off again in the densest areas, rather than climbing all the way up.

Reviewer 3: Is "Lowering it bends the noise into a bell" meant as "Lowering gives the grain a bell shape"?

density floor
: The density each crystal sits at even where the film received no light -- the reason grain does not disappear entirely in clear areas. Real stocks measure between about 0.03 and 0.06.

### texture

grain blur
: How much the grain is blurred after it is drawn, in pixels. This softens grain rather than resizing it: the crystals are unchanged, so raising it makes grain smoother and less distinct, not coarser. Use _granularity_ above for that.
: The value that looks right depends on output size, since it is measured in pixels rather than on the film. Raise it for large prints where one crystal covers several pixels; lower it for small output.

dye cloud size
: Each developed crystal leaves a small cloud of dye rather than a hard dot. This scales how far that cloud spreads, measured on the film itself, so it only becomes visible at magnifications where a single crystal covers more than a pixel.

recovery radius

> : How wide the recovery sharpening reaches, in pixels. 0 switches it off. A broad radius puts the crispness back into mid-sized detail rather than at the pixel level, which is what stops the pixel grid itself from showing through.

Reviewer 3: ```suggestion
: How wide the recovery sharpening reaches, in pixels. 0 switches it off. A broad radius puts the crispness back into mid-sized detail. This stops the pixel grid itself from showing through.
```

recovery strength
: How strongly it sharpens, restoring the definition the grain blur took away. It moves density around without adding any, so it cannot invent detail and is not a noise reduction. 0 leaves the softening in place with nothing countering it; pushing well beyond the default sharpens more than was ever lost, which makes grain look crunchy rather than photographic.

## halation

> Two separate things happen to bright light inside film. Some of it scatters sideways within the emulsion as it passes through. Some goes all the way through, reflects off the film base behind, and comes back -- which is what puts a reddish glow around very bright highlights. Both are simulated, and adjusted independently.

Reviewer 3: ```suggestion
Two separate things happen to bright light inside film. Some of it scatters sideways within the emulsion as it passes through. Some goes all the way through, reflects off the film base behind, comes back and puts a reddish glow around very bright highlights. Both are simulated, and adjusted independently.
```

enable halation
: Switch both effects on.

### scatter

scatter amount

> : How much light scatters sideways within the emulsion. 1.0 matches the film, 0 switches it off. This is a fraction of the light, so 1.0 -- all of it -- is genuinely the maximum, unlike _halation strength_ below.

Reviewer 3: I don't get the second sentence. The "all of it", the "genuinely", the whole idea.

scatter size
: How far it scatters. 1.0 matches the film and is the value the simulation normally works at. Above that the whole frame softens quickly, since the radius grows directly with the value. Drags to 1.5, right-click for more.

### halation

halation strength

> : How strong the reflected glow is. 1.0 matches the chosen film, which means it looks very different from stock to stock: most modern colour negative film has a layer specifically to absorb this light and shows almost none, while film without one (a redscale-style stock, say) glows dramatically. That difference belongs to the real film, it is not an inconsistency. Drags to 2, right-click for up to 8.

Reviewer 3: What is meant by "That difference belongs to the real film, it is not an inconsistency."

halation size
: How far the glow spreads. 1.0 matches the film.

### threshold

highlight boost

> : Rebuilds highlights that were clipped in the original file, so they can glow into the scatter, halation and diffusion effects instead of sitting flat. In EV; 0 switches it off. The boost applies over a fixed range, so it gives the same result regardless of image size, zoom level, or how the export was processed.

Reviewer 3: "X instead of"

boost range
: How far down the tonal range the boost reaches. Lower keeps it to the very brightest areas; higher pulls more of the upper midtones into the glow.

boost protect
: Keeps everything below this many stops above mid-grey out of the boost entirely.

## diffusion

Simulates a diffusion filter -- the physical piece of glass a cinematographer puts in front of the lens to bloom highlights and take the edge off contrast. There are two here, at different points in the chain: one at the camera, one at the enlarger.

enable diffusion filter
: Switch on the camera-stage filter.

### film

diffusion filter type
: Which filter to imitate:
: - _black pro-mist_: a concentrated, punchy halo that keeps blacks deep.
: - _glimmerglass_: tight and subtle, preserving sharpness.
: - _pro-mist_: broader and more pastel, a softer atmospheric look.
: - _cinebloom_: a wide, slowly fading veil across the frame.

diffusion strength
: How much light is diverted into the halo. 0 switches it off. The halo is added on top of the unfiltered picture, so raising this lifts shadows and lowers contrast as well as making highlights glow.

diffusion size

> : How far the halo spreads -- the same light carried further, rather than more of it. Use _diffusion strength_ for more.

Reviewer 3: "X rather than Y"
